File: selenium-java/src/test/java/pages/RegistroPage.java
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.Select;

/**
 * Page Object de la pagina de inscripcion de votantes.
 *
 * La prueba habla el lenguaje del negocio ("inscribir un votante") y esta
 * clase traduce eso a interacciones con el navegador. Si manana cambia el
 * identificador de un campo, se corrige AQUI y ninguna prueba se entera.
 *
 * Sobre los localizadores: se usan ids, que son el localizador mas estable.
 * Se evitan a proposito los XPath absolutos (/html/body/div[2]/form/div[3])
 * y las clases de CSS generadas, porque se rompen con cualquier rediseno.
 */
public class RegistroPage extends BasePage {

    private static final By NOMBRE = By.id("nombre");
    private static final By DOCUMENTO = By.id("documento");
    private static final By EDAD = By.id("edad");
    private static final By GENERO = By.id("genero");
    private static final By VIVO = By.id("vivo");
    private static final By BOTON_REGISTRAR = By.id("btn-registrar");
    private static final By RESULTADO_TITULO = By.id("resultado-titulo");
    private static final By ERROR_DOCUMENTO = By.id("error-documento");

    private final String baseUrl;

    public RegistroPage(WebDriver driver, String baseUrl) {
        super(driver);
        this.baseUrl = baseUrl;
    }

    public RegistroPage abrir() {
        driver.get(baseUrl + "/");
        waitForVisible(NOMBRE);
        return this;
    }

    public RegistroPage escribirNombre(String nombre) {
        type(NOMBRE, nombre);
        return this;
    }

    public RegistroPage escribirDocumento(int documento) {
        type(DOCUMENTO, String.valueOf(documento));
        return this;
    }

    public RegistroPage escribirEdad(int edad) {
        type(EDAD, String.valueOf(edad));
        return this;
    }

    public RegistroPage seleccionarGenero(String genero) {
        new Select(waitForVisible(GENERO)).selectByValue(genero);
        return this;
    }

    /** Marca o desmarca la casilla "La persona esta viva". */
    public RegistroPage marcarViva(boolean viva) {
        boolean actual = waitForVisible(VIVO).isSelected();
        if (actual != viva) {
            click(VIVO);
        }
        return this;
    }

    public RegistroPage registrar() {
        click(BOTON_REGISTRAR);
        return this;
    }

    /** Titulo del mensaje de resultado, esperando a que aparezca. */
    public String resultado() {
        return getText(RESULTADO_TITULO);
    }

    /** Mensaje de error del campo documento. */
    public String errorDocumento() {
        return getText(ERROR_DOCUMENTO);
    }

    /**
     * Inscribe a una persona en un solo paso.
     * Las pruebas que no necesitan controlar cada campo usan este atajo.
     */
    public RegistroPage inscribir(String nombre, int documento, int edad, String genero, boolean viva) {
        return escribirNombre(nombre)
                .escribirDocumento(documento)
                .escribirEdad(edad)
                .seleccionarGenero(genero)
                .marcarViva(viva)
                .registrar();
    }

    /** Genera un documento distinto en cada llamada. */
    public static int documentoUnico() {
        return (int) (Math.random() * 900000) + 100000;
    }
}

File: playwright/pages/RegistroPage.js
// @ts-check
const { expect } = require('@playwright/test');

/**
 * PAGE OBJECT MODEL de la pagina de inscripcion.
 *
 * La idea: la prueba habla el lenguaje del NEGOCIO ("inscribir un votante"),
 * y esta clase traduce eso a interacciones con el navegador.
 *
 * El beneficio real aparece cuando la interfaz cambia. Si el boton cambia de
 * texto, se corrige AQUI, en un solo lugar, y las 8 pruebas que lo usan
 * siguen funcionando sin tocarse. Sin POM habria que editar 8 archivos.
 */
class RegistroPage {

  constructor(page) {
    this.page = page;

    // Localizadores por ROL y texto visible: sobreviven a un rediseño y de
    // paso comprueban que el elemento es accesible.
    this.nombre = page.getByLabel('Nombre completo');
    this.documento = page.getByLabel('Número de documento');
    this.edad = page.getByLabel('Edad');
    this.genero = page.getByLabel('Género');
    this.vivo = page.getByLabel('La persona está viva');
    this.botonRegistrar = page.getByRole('button', { name: 'Registrar votante' });
    this.resultado = page.getByRole('status');
  }

  async abrir() {
    await this.page.goto('/');
  }

  /**
   * Inscribe a una persona. Los campos no indicados usan valores por defecto
   * razonables, para que cada prueba solo declare lo que le importa.
   */
  async inscribir({ nombre = 'Persona Prueba', documento, edad = 30, genero = 'FEMALE', vivo = true }) {
    await this.nombre.fill(nombre);
    await this.documento.fill(String(documento));
    await this.edad.fill(String(edad));
    await this.genero.selectOption(genero);
    if (vivo) {
      await this.vivo.check();
    } else {
      await this.vivo.uncheck();
    }
    await this.botonRegistrar.click();
  }

  /** Verifica el titulo del mensaje de resultado. */
  async esperarResultado(titulo) {
    await expect(this.page.getByRole('heading', { name: titulo })).toBeVisible();
  }

  /** Genera un documento distinto en cada llamada. */
  static documentoUnico() {
    return Math.floor(Math.random() * 900000) + 100000;
  }
}

module.exports = { RegistroPage };
